### ISSUE TITLE
Playback issue when phone sleeps (i.e., turn display off)

### ISSUE THREAD

**Issue by The3Designer** [OxygenCobalt/Auxio#151]
#### Describe the bug/crash:
When turning off my phone's display by pushing the power button once (i.e., putting the device in sleep mode), the speed of the playback is reduced (I guess about 0.3 the speed of normal) and the playback stutters/gets distorted. As soon as I push the power button again, which brings up the lock screen, the playback resumes to normal. The same happens when I scan my finger from sleep, which brings up Android.

This issue did not occur with the Metro app (link: https://f-droid.org/en/packages/io.github.muntashirakon.Music/), which makes me guess it's an issue with this app.

I've tried enabling and disabling ReplayGain, this makes no difference.
I've tried enabling and disabling battery optimization for this app, this makes no difference.
I've tried moving the app from external storage to internal storage, this makes no difference.
I've tried turning off the screen with the app open or the app running in the background, this makes no difference.

My audio library is about 350 FLAC files.

**This issue only occurs on certain songs!**
One song, 125 MB with 2672 kbps bitrate and 88200 sample rate, has no issues when playing while phone is sleeping.
Another song, 92 MB with 2866 kbps bitrate and 96000 sample rate, does have issues when playing while phone is sleeping.

Could this be a bitrate and/or sample rate issue? If so, how could we fix this, given that the aforementioned Metro app had no issues playing any file/song even with the display off.
_____________________________________________________________________________________________________________________

#### Expected behavior
Auxio should keep playing the music at the normal speed, without choppy or buggy playback.
_____________________________________________________________________________________________________________________

#### Steps To Reproduce the bug/crash:
1. Open Auxio
2. Play a song
3. Push the phone's power button once (screen goes off)
4. Music playback slows down and stutters
5. Push power button once again (lock screen appears)
6. Music starts playing normal again

So it doesn't matter that the phone is locked, but as soon as the screen/display goes off (i.e., phone goes to sleep), Auxio's playback slows down and gets choppy/distorted.
_____________________________________________________________________________________________________________________

#### Logs/Stack Traces:
Sorry, don't have this.
_____________________________________________________________________________________________________________________

#### Screenshots:
n.a.
_____________________________________________________________________________________________________________________

#### Phone Information:
Xiaomi Redmi Note 3 Pro running Android 11 via ColtOS 2021 ENIGMA custom ROM (link to ROM: https://forum.xda-developers.com/t/rom-unofficial-coltos-7-4-kenzo-kate-11-ota-enforcing-mar-30-2022.4090359/).

I've downloaded Auxio from the F-Droid store.
_____________________________________________________________________________________________________________________

#### Due Diligence:
- [x] I have checked this issue for any duplicates.
- [x] I have checked for this issue in the [FAQ](https://github.com/OxygenCobalt/Auxio/blob/dev/info/FAQ.md).
- [x] I have read the [Contribution Guidelines](https://github.com/OxygenCobalt/Auxio/blob/dev/.github/CONTRIBUTING.md).

**Comment by OxygenCobalt** [comment]
Thank you so much for ruling out other cases. I would say that this is one of two issues:

1. 2866 kbps bitrate and 96000 sample rate is *huge.* This might feasibly bog down ExoPlayer, especially on a budget brand like Xiaomi and in the low-power mode that Auxio is put in when the phone is asleep. For comparison, Metro uses MediaPlayer, which while native is much less capable of a good user experience.
2. Even if ReplayGain is off, I still have to copy the audio buffer due to how I handle the feature. This might become a bottleneck given your large bitrate.

Can you try an older, pre-ReplayGain version of Auxio (see below, rename the extension from ZIP) and see if the issue persists? That way I can rule out an issue on my end.

[Auxio_2.0.1.zip](https://github.com/OxygenCobalt/Auxio/files/8884033/Auxio_2.0.1.zip)

**Comment by The3Designer** [comment]
Thanks for your fast reply @OxygenCobalt !

I tried your 2.0.1. version, and this version does not have the issue! I even played a file with 5154 kbps bitrate and 192000 sample rate, and it did not stutter when turning the display off!

So I guess it's the ReplayGain implementation then that causes the issue, even if it is turned off in the settings?

**Comment by OxygenCobalt** [comment]
Yeah, I guess it is. This is not good, as I *cannot disable the replaygain system once it's enabled.* I need to make my implementation more efficient to resolve this issue.

Can you try this APK? I cannot optimize what occurs when ReplayGain is enabled any further, but I can optimize what occurs when it's disabled. Note that this is technically a release 2.3.1 build, so it will update your existing app.

[Auxio_FasterCopy.zip](https://github.com/OxygenCobalt/Auxio/files/8884173/Auxio_FasterCopy.zip)

**Comment by The3Designer** [comment]
That version of the app does have the problem. I uninstalled the old version, then installed your modified 2.3.1 version. Immediatly after the install I started playing music (i.e., ReplayGain was off and never enabled). Some songs did have the issue, others did not, similar to how I originally reported the issue.

**Comment by OxygenCobalt** [comment]
I don't think there is much I can do about this. Even with the optimizations, it looks like your device is too under-powered and your music files are too resource-intensive for it to feasibly handle it.

There are two options:
- In the mean time, I would recommend maintaining an Auxio build that disables ReplayGain *entirely.* You can do this by going to [this function](https://github.com/OxygenCobalt/Auxio/blob/d42f5970f1dd24ed6772a1d554968ff205dcac20/app/src/main/java/org/oxycblt/auxio/playback/system/PlaybackService.kt#L326) and removing all instances of `replayGainProcessor`, and then following the typical steps to build the app (Including [Auxio's build steps](https://github.com/OxygenCobalt/Auxio#building)
- In the future, I might do what someone requested in [this issue for another app](https://github.com/enricocid/Music-Player-GO/issues/413) and add some kind of "loss-less audio" option. In such a mode, the ReplayGain system would not be enabled, and thus your issue would be solved.

If you don't want to build the app yourself, you can use this APK. Let me know if this solves your issue.

[Auxio_NoReplayGain.zip](https://github.com/OxygenCobalt/Auxio/files/8884210/Auxio_NoReplayGain.zip)

**Comment by The3Designer** [comment]
Sadly, the Auxio_NoReplayGain.apk versions still has the issue on some files. I am not able to build the app myself, so for now I will switch back to the Metro app.

Thanks for your effort and quick communication so far! Hopefully this can be fixed in the future, perhaps through a "lossless audio" option :)

**Comment by OxygenCobalt** [comment]
Okay. I'll close this issue as a wontfix then. Super sorry about this. I guess this is the cost of using ExoPlayer. While it's more extendable than `MediaPlayer`, the added overhead is costly for audiophiles like you.

**Comment by The3Designer** [comment]
No problem, it's good to know the technical limitations of the backend and how this affects the app's functionality and user experience.

Thanks for your support and quick assistance! :)

**Comment by OxygenCobalt** [comment]
I just realized there might be another bottleneck. ExoPlayer actually downsamples high-quality audio into 16-bit PCM in normal operation. I'd imagine this adds *even more* overhead when playing audio.

Luckily, I can disable it with that lossless audio idea that I brought up prior. Only issue is that doing such swaps out *one* resampling system with *another.* Yeah. I think an implementation of that "Lossless audio" option is clearer now, but it once again requires me to write my own ExoPlayer components.

**Comment by OxygenCobalt** [comment]
Also, better still, *I found a way to disable the ReplayGain audio processor*. Turns out I need to switch to the lower-level audio processor class and operate with that.

**Comment by OxygenCobalt** [comment]
@The3Designer I have some more updates.

- I was able to make the ReplayGain audio processor disable itself dynamically, allowing me to skip a copy
- After doing some communication with the ExoPlayer devs, I'm going to try to start work on a pull request that should remove most of the resampling done in the audio sink.
- In the future, resampling should then only be done on playback speed adjustment (Which Auxio does not have) and if the source format is not supported by the OS. I think this will remove some of the overhead that you need to deal with. 

Assuming that your problem FLAC files are 24-bit or 32-bit, there should only be one copy from source to output on Android 12, and one resampling pass done (24/32-bit PCM -> Float) on versions below 12. This is still a massive improvement from previously.

**Comment by The3Designer** [comment]
@OxygenCobalt Thanks for the updates on this topic, and sorry for my lack of response.

It's good to hear that progress is being made (also in collaboration with the ExoPlayer devs). Just a quick update from my side: I've recently upgraded to a different phone, for which I've converted my FLAC library to 320 kbps .mp3 files (to reduce the library size). With this more powerful smartphone and the greatly reduced filesize (and other relevant specs like bitrate, sampling rate, etc.) I can play songs with the screen off without any issues.

When the new (re)sampling code/style is implemented into an APK, I'd be happy to test it again with the more demanding FLAC files :)

For now, thanks for all your effort and the great app!
